### ISSUE TITLE
kvserver: finished proposal inserted into map

### ISSUE THREAD

**Issue by tbg** [cockroachdb/cockroach#97605]
**Describe the problem**

Seen in https://github.com/cockroachdb/cockroach/issues/96149 and also in TestSQLStatsRegions[^1]

The assertion was introduced relatively recently[^2] so this is likely not a regression.

[^1]: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/8811851?showRootCauses=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true
[^2]: 15b1c6ae6058c8060eb8afe2110e582106380e96

Jira issue: CRDB-24780

Epic CRDB-25287

**Comment by tbg** [comment]
The proposal was disabled in https://github.com/cockroachdb/cockroach/pull/97606.

Some musings on this.

The ownership model of proposals (`*ProposalData`) is quite complex. Proposals live in the `r.mu.proposals` map, but "at any time" they could be up for application, in which case there'd be access to them below raft (and they'd get removed from the map). "Safe" access to a `*ProposalData` (once it's in the proposals map) requires holding `raftMu`; the map itself is under `repl.mu` so you end up holding both. At the same time, a proposal can refer to multiple inflight log entries (`tryReproposeWithNewLeaseIndex`). If we called `tryReproposeWithNewLeaseIndex` during application of an entry, we've temporarily taken it out of the map and it's now sitting in the proposal buffer anew.

On to the proposal buffer. We hold `r.mu` (and `raftMu`) when inserting into the buffer, then we release both mutexes. The proposal is sitting in a slice[^1] after that point. (It may also still be in the proposal map). It won't be touched until the buffer is flushed, though, at which point both mutexes are held again.

I'm not sure what went wrong in https://github.com/cockroachdb/cockroach/issues/96149, but there is a lot going on here.

[^1]: https://github.com/cockroachdb/cockroach/blob/4eb5451dcfbac79906eb48340f76597ca4d8ead2/pkg/kv/kvserver/replica_proposal_buf.go#L267
[^2]: https://github.com/cockroachdb/cockroach/blob/a363143b81f5045deab1ab8073da6d6949984936/pkg/kv/kvserver/replica_raft.go#L1313

**Comment by tbg** [comment]
`(*Replica).tick`[^1] does *not* hold `raftMu`. So you can have the following interleaving.

- proposal inserted into map & proposal buffer
- proposal is replicated and just about to come up for application (but not yet!)
- proposal is resubmitted to the proposal buffer via `tick`
- application starts, command is accepted (proposal is finished)
- but there is a copy sitting in the proposal buffer still!
- when the proposal buffer next flushes (i.e. an instant later) the assertion fires.

[^1]: https://github.com/cockroachdb/cockroach/blob/a363143b81f5045deab1ab8073da6d6949984936/pkg/kv/kvserver/replica_raft.go#L1313-L1434

**Comment by tbg** [comment]
There are some more dragons lurking here. When flushed from the propbuf, the command may also be rejected:

https://github.com/cockroachdb/cockroach/blob/4eb5451dcfbac79906eb48340f76597ca4d8ead2/pkg/kv/kvserver/replica_proposal_buf.go#L429

we'd then be doubly-finishing a proposal. This won't crash (it becomes a noop the second time around) but it smells that this is possible.

There are some replica circuit breaker related crashes that I had looked into (on a plane, so can't link it right now) that seem related. They can probably be explained by variants of the above - the crash was caused by trying to poison a request's latches, but finding that request already finished. I had not understood why a finished request could be in the proposals map - that's precisely why I added the assertion[^1] - and now it makes more sense. In fact, the circuit breaker crash should now be replaced by the assertion crash discussed in this issue.

[^1]: 15b1c6ae6058c8060eb8afe2110e582106380e96
